### PR TITLE
[WIP] Adding new Download verb

### PR DESCRIPTION
**Pull request by gsegatti**
This is an initial attempt to solve issue #1760.

The idea here is to provide the minimal set of changes which would permit to populate our cache with a new verb. This way we can (in the near future) implement taking mkosi manifestos and so on as explained on the issue.

### CODE REVIEW

File: mkosi/__init__.py [hunk 1]
```diff
@@ -2170,6 +2173,14 @@ def run_verb(args: MkosiArgs, presets: Sequence[MkosiConfig]) -> None:
 
     last = presets[-1]
 
+    if args.verb == Verb.download:
+        workspace = tempfile.TemporaryDirectory(dir=last.workspace_dir, prefix=".mkosi-tmp")
```
Comment (gsegatti) [inline mkosi/__init__.py:2177]: I'm necessarily using `last` here, for the sake of simplicity while testing. But from my understanding we can receive a list (or `Sequence`) of `MkosiConfig`.

Each config can carry it's own distribution. So perhaps we should, for each `preset`, download build cache of its distribution?

File: mkosi/__init__.py [hunk 2]
```diff
@@ -2170,6 +2173,14 @@ def run_verb(args: MkosiArgs, presets: Sequence[MkosiConfig]) -> None:
...
+        with workspace, scopedenv({"TMPDIR" : workspace.name}):
```
Comment (gsegatti) [inline mkosi/__init__.py:2178]: Also this `with` statement is duplicated here and in `build_images()`. Maybe we should opt for a function with the context manager decorator for this (assuming this is even the correct way to go)?

**Comment by DaanDeMeyer** [comment]
Each package manager should have a way to just download packages without installing them. That's what we should use to implement this functionality. Also this should be done for each preset. So loop over each preset and download the specified packages and build packages. And then later we can extend that by passing in the packages from the manifest to download specific versions of packages.

**Comment by gsegatti** [comment]
Hey Daan!
> Also this should be done for each preset. So loop over each preset and download the specified packages and build packages. 

Appreciate you clarifying this, I thought so.

> Each package manager should have a way to just download packages without installing them.

I see, I tracked this down to the `DistributionInstaller` class for each distribution. Do you reckon we should implement a new Class Method `download_packages()` or should adapt one of the existing methods? I think a new one would be nice for readability/cleanliness, but let me know what you guys think.

File: mkosi/__init__.py (hunk 1 above)
Comment (gsegatti) [reply mkosi/__init__.py:2177]: Marking this as resolved since Daan clarified we should loop over all presets.

**Comment by DaanDeMeyer** [comment]
A new method seems like the way to go I think.

**Comment by DaanDeMeyer** [comment]
We should probably have manifest support from the start in this because I'm not entirely sure whether the way we currently do things would work with the distribution provided archives to download older packages from. For Arch Linux we might need to do things quite differently than how we usually do things.

**Comment by gsegatti** [comment]
It's been sometime since I last touched this, I was in the middle of moving to a different city. If what I've done so far isn't what we hoped towards implementing the `Download` verb, maybe it is best to close this PR for now?

**Comment by DaanDeMeyer** [comment]
Frankly I'm not entirely sure myself, it might be different depending on the distribution. Let's close this one for now, but feel free to open a new PR if you want

**Pull request closed without merging**